### PR TITLE
Add parquet as a writer / datasource type to reth-indexer

### PR DESCRIPTION
**Pull request by liangjh**
Add ability to write to local parquet file(s) within the existing reth-indexer multi-datasource framework. 

Utilize polars dataframe construction and serialization to fast-write to parquet format.

Generalize some type and column name components written for GCP to be utilized in parquet writer (+ can be used in other writers / formats in future). 

Linked to issue: https://github.com/joshstevens19/reth-indexer/issues/38

### CODE REVIEW

File: src/datasource.rs [hunk 1]
```diff
@@ -1,12 +1,133 @@
...
+                    )
+                })
+                .collect();
+
+            let common_column_type_map: IndexMap<String, String> = COMMON_COLUMNS
+                .into_iter()
+                .map(|it| (it.0.to_string(), it.1.to_string()))
+                .collect();
+
+            let merged_column_types: IndexMap<String, String> = common_column_type_map
```
Comment (joshstevens19) [inline src/datasource.rs:66]: you can most likely chain these into 1 let with `common_column_type_map`

Comment (joshstevens19) [inline src/datasource.rs:66]: you can most likely chain these into 1 let with `common_column_type_map`

File: src/parquet.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+/// * `event_mappings` - list of events and abi fields to parse
+pub async fn init_parquet_db(
+    indexer_parquet_config: &IndexerParquetConfig,
+    event_mappings: &[IndexerContractMapping],
+) -> Result<ParquetClient, ParquetClientError> {
+    let parquet_client = ParquetClient::new(indexer_parquet_config, event_mappings).await?;
+
+    if indexer_parquet_config.drop_tables {
+        let res = parquet_client.delete_tables().await;
+        match res {
```
Comment (joshstevens19) [inline src/parquet.rs:37]: can you use an If let Error( instead of match here?

File: src/parquet.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                return Err(ParquetClientError::ParquetOperationError(format!(
+                    "Parquet client operation error: {:?}",
+                    err
+                )))
+            }
+        }
+    }
+
+    let res = parquet_client.create_tables().await;
+    match res {
```
Comment (joshstevens19) [inline src/parquet.rs:49]: same as above

File: src/parquet.rs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        Ok(..) => {}
+        Err(err) => {
+            return Err(ParquetClientError::ParquetOperationError(format!(
```
Comment (joshstevens19) [inline src/parquet.rs:52]: probs but this in a fn and share it as its pasted 3 times in here

File: src/parquet.rs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        indexer_parquet_config: &IndexerParquetConfig,
+        indexer_contract_mappings: &[IndexerContractMapping],
+    ) -> Result<Self, ParquetClientError> {
+        let table_map = load_table_configs(indexer_contract_mappings);
+
+        // Check data directory exists - if not, then create!
+        let root_data_dir: &Path = Path::new(indexer_parquet_config.data_directory.as_str());
+        if !root_data_dir.exists() {
+            let res = fs::create_dir(root_data_dir);
+            match res {
```
Comment (joshstevens19) [inline src/parquet.rs:81]: same as above

File: src/parquet.rs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    ///
+    pub async fn delete_tables(&self) -> Result<(), Box<dyn Error>> {
+        if self.drop_tables {
+            for table_name in self.table_map.keys() {
+                let table_dir = format!("{}/{}/", self.data_directory, table_name);
+                let table_path = Path::new(table_dir.as_str());
+                if table_path.exists() && table_path.is_dir() {
+                    println!("Removing directory / contents: {}", table_dir);
+                    let res = fs::remove_dir_all(table_path);
+                    match res {
```
Comment (joshstevens19) [inline src/parquet.rs:111]: same as a bove

File: src/parquet.rs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    ///
+    pub async fn create_tables(&self) -> Result<(), Box<dyn Error>> {
+        for table_name in self.table_map.keys() {
+            // Check if table exists;  if NO, then create new
+            let table_dir = format!("{}/{}/", self.data_directory, table_name);
+            let table_path = Path::new(table_dir.as_str());
+            if !table_path.exists() {
+                println!("Creating directory: {}", table_name);
+                let res = fs::create_dir(table_path);
+                match res {
```
Comment (joshstevens19) [inline src/parquet.rs:134]: same as above

**Review by joshstevens19 — CHANGES_REQUESTED** [review]
tiny tiny code changes but over nice PR

**Comment by liangjh** [comment]
thx for the feedback re: `if let Err(...` pattern.  updated, lmk what you think -

**Review by joshstevens19 — APPROVED** [review]

**Pull request merged**